File: solutions/ocaml/055-design-in-memory-file-system.ml
(*
  Problem 55: Design In-Memory File System (LeetCode 588)
  Difficulty: Hard
  Language: OCaml
*)
module SMap = Map.Make(String)

type fs_node = { children: fs_node SMap.t; content: string }

let empty_node = { children = SMap.empty; content = "" }

let navigate root parts =
  List.fold_left (fun node p ->
    try SMap.find p node.children
    with Not_found -> empty_node
  ) root parts

let rec ensure_path node = function
  | [] -> node
  | p :: ps ->
    let child = try SMap.find p node.children with Not_found -> empty_node in
    { node with children = SMap.add p (ensure_path child ps) node.children }

let rec update_at node parts f = match parts with
  | [p] ->
    let child = try SMap.find p node.children with Not_found -> empty_node in
    { node with children = SMap.add p (f child) node.children }
  | p :: ps ->
    let child = try SMap.find p node.children with Not_found -> empty_node in
    { node with children = SMap.add p (update_at child ps f) node.children }
  | [] -> f node

let parse_path path =
  String.split_on_char '/' path |> List.filter (fun s -> s <> "")

let ls root path =
  let parts = parse_path path in
  let node = navigate root parts in
  if node.content <> "" then [List.nth parts (List.length parts - 1)]
  else SMap.bindings node.children |> List.map fst

File: solutions/clojure/055-design-in-memory-file-system.clj
;; Problem 55: Design In-Memory File System (LeetCode 588)
;; Difficulty: Hard
;; Language: Clojure
;; 
(defn mk-node [] {:children {} :content ""})

(defn parse-path [path]
  (filterv (complement empty?) (clojure.string/split path #"/")))

(defn navigate [root parts]
  (reduce (fn [node p]
            (get-in node [:children p] (mk-node)))
          root parts))

(defn ensure-path [root parts]
  (if (empty? parts) root
    (let [[p & ps] parts
          child (get-in root [:children p] (mk-node))]
      (assoc-in root [:children p] (ensure-path child ps)))))

(defn fs-ls [root path]
  (let [parts (parse-path path)
        node (navigate root parts)]
    (if (seq (:content node))
      [(peek parts)]
      (sort (keys (:children node))))))

(defn fs-mkdir [root path]
  (ensure-path root (parse-path path)))

(defn fs-add-content [root path content]
  (let [parts (parse-path path)]
    (update-in (ensure-path root parts)
               (into [] (interleave (repeat :children) parts))
               (fn [node] (update node :content str content)))))
